File: r/tests/testthat/test-dw_pop.R
# Regression tests for dw_pop() (#17).
#
# Strategy: drop a fixture CSV at the cache path the helper looks up,
# then exercise the latest-year, year-filter, and country-filter branches
# without hitting the network.

.dw_pop_fixture <- function() {
  # Three countries x three years, with one missing-row case so the
  # complete-case filter is exercised.
  data.frame(
    iso3c        = c("BFA", "BFA", "BFA",
                     "MLI", "MLI", "MLI",
                     "NER", "NER", "NER",
                     "XXX"),
    country      = c("Burkina Faso", "Burkina Faso", "Burkina Faso",
                     "Mali", "Mali", "Mali",
                     "Niger", "Niger", "Niger",
                     "Bogus"),
    date         = c(2021, 2022, 2023,
                     2021, 2022, 2023,
                     2021, 2022, 2023,
                     2023),
    value        = c(22.1e6, 22.6e6, 23.2e6,
                     21.9e6, 22.6e6, 23.3e6,
                     25.1e6, 26.2e6, 27.2e6,
                     NA),
    stringsAsFactors = FALSE
  )
}

test_that("dw_pop returns latest year per country by default", {
  d <- local_tempdir()
  cache_dir <- file.path(d, "_apis", "wb")
  dir.create(cache_dir, recursive = TRUE)
  write.csv(.dw_pop_fixture(),
            file = file.path(cache_dir, "wb_population_sp_pop_totl.csv"),
            row.names = FALSE)
  local_state(teamsRawData = d, teamsRawDataCanonical = d,
              dw_apis_allowed = FALSE)

  out <- dw_pop()
  expect_s3_class(out, "data.frame")
  expect_equal(names(out), c("REF_AREA", "TIME_PERIOD", "OBS_VALUE"))
  # Three countries with valid rows; bogus XXX has NA OBS_VALUE -> dropped
  expect_equal(nrow(out), 3L)
  # Every row should be from 2023 (latest)
  expect_true(all(out$TIME_PERIOD == 2023L))
  expect_setequal(out$REF_AREA, c("BFA", "MLI", "NER"))
})

test_that("dw_pop with year=2022 returns one year per country", {
  d <- local_tempdir()
  cache_dir <- file.path(d, "_apis", "wb")
  dir.create(cache_dir, recursive = TRUE)
  write.csv(.dw_pop_fixture(),
            file = file.path(cache_dir, "wb_population_sp_pop_totl.csv"),
            row.names = FALSE)
  local_state(teamsRawData = d, teamsRawDataCanonical = d,
              dw_apis_allowed = FALSE)

  out <- dw_pop(year = 2022)
  expect_equal(nrow(out), 3L)
  expect_true(all(out$TIME_PERIOD == 2022L))
})

test_that("dw_pop with countries filter returns only the named ISO3 codes", {
  d <- local_tempdir()
  cache_dir <- file.path(d, "_apis", "wb")
  dir.create(cache_dir, recursive = TRUE)
  write.csv(.dw_pop_fixture(),
            file = file.path(cache_dir, "wb_population_sp_pop_totl.csv"),
            row.names = FALSE)
  local_state(teamsRawData = d, teamsRawDataCanonical = d,
              dw_apis_allowed = FALSE)

  out <- dw_pop(countries = c("BFA", "MLI"))
  expect_setequal(out$REF_AREA, c("BFA", "MLI"))
})

test_that("dw_pop reviewer-mode with empty cache raises the envelope", {
  d <- local_tempdir()
  local_state(teamsRawData = d, teamsRawDataCanonical = d,
              dw_apis_allowed = FALSE,
              dw_mode = "reviewer")

  err <- tryCatch(
    dw_pop(cache_key = "never_cached_pop_key"),
    error = identity
  )
  expect_s3_class(err, "error")
  # `dw_require_no_api()` raises the envelope; we only assert that the
  # call fails clearly.  The exact prefix depends on the consumer's
  # profile setup.
})

test_that("dw_pop sorts output by REF_AREA + TIME_PERIOD ascending", {
  d <- local_tempdir()
  cache_dir <- file.path(d, "_apis", "wb")
  dir.create(cache_dir, recursive = TRUE)
  write.csv(.dw_pop_fixture(),
            file = file.path(cache_dir, "wb_population_sp_pop_totl.csv"),
            row.names = FALSE)
  local_state(teamsRawData = d, teamsRawDataCanonical = d,
              dw_apis_allowed = FALSE)

  out <- dw_pop(year = c(2021, 2022, 2023))
  # 3 countries x 3 years = 9 rows
  expect_equal(nrow(out), 9L)
  expect_equal(out$REF_AREA, sort(out$REF_AREA))
})

File: python/tests/test_dw_pop.py
"""Regression tests for ``dw_pop`` (#17).

Python mirror of ``r/tests/testthat/test-dw_pop.R``.

Strategy: drop a fixture CSV at the cache path the helper looks up, then
exercise the latest-year, year-filter, and country-filter branches
without hitting the network.

Run with::

    pytest python/tests/test_dw_pop.py
"""

from __future__ import annotations

import shutil
import sys
import tempfile
from pathlib import Path

import pandas as pd
import pytest


def _bootstrap_package() -> None:
    """Make ``python/src/`` importable under the name ``cso_toolkit``."""
    try:
        import cso_toolkit  # noqa: F401
        return
    except ImportError:
        pass
    repo_root = Path(__file__).resolve().parents[2]
    src = repo_root / "python" / "src"
    tmpdir = Path(tempfile.mkdtemp(prefix="cso_toolkit_pytest_"))
    pkg_dir = tmpdir / "cso_toolkit"
    shutil.copytree(src, pkg_dir)
    sys.path.insert(0, str(tmpdir))


_bootstrap_package()

from cso_toolkit import _state, dw_pop  # noqa: E402


def _fixture() -> pd.DataFrame:
    """Three countries x three years, plus one NA-value bogus row so the
    complete-case filter is exercised."""
    return pd.DataFrame({
        "iso3c": ["BFA", "BFA", "BFA",
                  "MLI", "MLI", "MLI",
                  "NER", "NER", "NER",
                  "XXX"],
        "country": ["Burkina Faso", "Burkina Faso", "Burkina Faso",
                    "Mali", "Mali", "Mali",
                    "Niger", "Niger", "Niger",
                    "Bogus"],
        "date": [2021, 2022, 2023,
                 2021, 2022, 2023,
                 2021, 2022, 2023,
                 2023],
        "value": [22.1e6, 22.6e6, 23.2e6,
                  21.9e6, 22.6e6, 23.3e6,
                  25.1e6, 26.2e6, 27.2e6,
                  float("nan")],
    })


@pytest.fixture()
def cached_state(tmp_path: Path):
    """Drop the fixture CSV at the WB cache path and configure state."""
    cache_dir = tmp_path / "_apis" / "wb"
    cache_dir.mkdir(parents=True)
    _fixture().to_csv(cache_dir / "wb_population_sp_pop_totl.csv", index=False)
    _state.configure(
        teamsRawData=str(tmp_path),
        teamsRawDataCanonical=str(tmp_path),
        dw_apis_allowed=False,
    )
    yield tmp_path


def test_latest_year_per_country(cached_state):
    out = dw_pop(verbose=False)
    assert list(out.columns) == ["REF_AREA", "TIME_PERIOD", "OBS_VALUE"]
    # Three countries with valid rows; bogus XXX has NA OBS_VALUE -> dropped.
    assert len(out) == 3
    assert (out["TIME_PERIOD"] == 2023).all()
    assert set(out["REF_AREA"]) == {"BFA", "MLI", "NER"}


def test_year_filter(cached_state):
    out = dw_pop(year=2022, verbose=False)
    assert len(out) == 3
    assert (out["TIME_PERIOD"] == 2022).all()


def test_countries_filter(cached_state):
    out = dw_pop(countries=["BFA", "MLI"], verbose=False)
    assert set(out["REF_AREA"]) == {"BFA", "MLI"}


def test_scalar_country_filter(cached_state):
    out = dw_pop(countries="NER", verbose=False)
    assert set(out["REF_AREA"]) == {"NER"}


def test_sorted_by_ref_area_then_time(cached_state):
    out = dw_pop(year=[2021, 2022, 2023], verbose=False)
    # 3 countries x 3 years = 9 rows.
    assert len(out) == 9
    assert list(out["REF_AREA"]) == sorted(out["REF_AREA"])


def test_reviewer_mode_empty_cache_raises(tmp_path):
    _state.configure(
        teamsRawData=str(tmp_path),
        teamsRawDataCanonical=str(tmp_path),
        dw_apis_allowed=False,
        dw_mode="reviewer",
    )
    with pytest.raises(PermissionError):
        dw_pop(cache_key="never_cached_pop_key", verbose=False)
